llvm_passes/HipPrintf.cpp

```diff
@@ -18,158 +53,363 @@
+    assert(OrigStr == nullptr || !OrigStr->hasInitializer());
+  }
+
+  ConstantDataSequential *CDSInitializer =
+      dyn_cast<ConstantDataSequential>(OrigStr->getInitializer());
+
+  assert(OrigStr->getInitializer()->isZeroValue() || CDSInitializer != nullptr);
 
-  Module *M = Before->getParent()->getParent()->getParent();
+  IsEmpty = OrigStr->getInitializer()->isZeroValue();
+  return (IsEmpty ? "" : std::string(CDSInitializer->getAsCString()));
+}
+
+// Tries to convert the string argument to an OpenCL compatible one
+// by creating a copy of it in the module global scope. Returns nullptr
+// if unable to do so.
+Value *HipPrintfToOpenCLPrintfPass::cloneStrArgToConstantAS(
+    Value *StrArg, llvm::IRBuilder<> &B, bool *IsEmpty) {
+
+  ConstantExpr *CE = dyn_cast<ConstantExpr>(StrArg);
+  if (CE == nullptr)
+    return nullptr;
 
   Type *Int8Ty = IntegerType::get(M->getContext(), 8);
-  ConstantExpr *CE = cast<ConstantExpr>(HipFmtStrArg);
 
-  Value *FmtStrOpr = CE->getOperand(0);
+  Value *StrOpr = CE->getOperand(0);
 
-  GlobalVariable *OrigFmtStr =
-    isa<GetElementPtrInst>(FmtStrOpr) ?
-    cast<GlobalVariable>(
-      cast<GetElementPtrInst>(FmtStrOpr)->getPointerOperand()) :
-    cast<GlobalVariable>(FmtStrOpr);
+  GlobalVariable *OrigStr = findGlobalStr(StrOpr);
+  if (OrigStr == nullptr || !OrigStr->hasInitializer())
+    return nullptr;
 
-  ConstantDataSequential *FmtStrData =
-    cast<ConstantDataSequential>(OrigFmtStr->getInitializer());
+  std::string NewStr = getCDSAsString(OrigStr, *IsEmpty);
+  return getOrCreateStrLiteralArg(NewStr, B);
+}
+
+// Checks if the GlobalVariable a literal string.
+bool isLiteralString(const GlobalVariable &Var) {
+
+  if (!Var.isConstant()) {
+    return false;
+  }
 
-  NumberOfFormatSpecs =
-    FmtStrData->getAsString().count("%") -
-    FmtStrData->getAsString().count("%%");
+  auto ArrayTy = dyn_cast<ArrayType>(Var.getType()->getElementType());
+  if (!ArrayTy) {
+    return false;
+  }
 
-  GlobalVariable *NewFmtStr = new GlobalVariable(
-      *M, OrigFmtStr->getValueType(), true, OrigFmtStr->getLinkage(),
-      FmtStrData, OrigFmtStr->getName() + ".cl",
-      (GlobalVariable *)nullptr, OrigFmtStr->getThreadLocalMode(),
-      SPIRV_OPENCL_PRINTF_FMT_ARG_AS);
-  NewFmtStr->copyAttributesFrom(OrigFmtStr);
+  auto IntTy = dyn_cast<IntegerType>(ArrayTy->getElementType());
+  if (!IntTy) {
+    return false;
+  }
+  return IntTy->getBitWidth() == 8;
+}
+
+// Create a new string literal that can be passed to an OpenCL printf
+// as an argument. Try to reuse previously created ones, if possible.
+Constant *
+HipPrintfToOpenCLPrintfPass::getOrCreateStrLiteralArg(const std::string &Str,
+                                                      llvm::IRBuilder<> &B) {
+
+  auto &LiteralArg = LiteralArgs[Str];
+  if (LiteralArg != nullptr)
+    return LiteralArg;
+
+  auto LiteralStr = B.CreateGlobalString(Str.c_str(), ".cl_printf_fmt_str",
```

Comment:
```suggestion
  auto *LiteralStr = B.CreateGlobalString(Str.c_str(), ".cl_printf_fmt_str",
```
(https://llvm.org/docs/CodingStandards.html#beware-unnecessary-copies-with-auto)